core/validation/validate.py

```diff
@@ -169,6 +170,36 @@ def validate_uniques(
     return non_unique_failures
 
 
+def validate_unique_togethers(
+    workbook: dict[str, pd.DataFrame], sheet_name: str, unique_togethers: list[tuple]
+) -> list[vf.NonUniqueTogetherFailure]:
+    """
+    Validate that compound keys' values in multiple columns are unique when together in a row.
+
+    :param workbook: A dictionary where keys are sheet names and values are
+                     pandas DataFrames.
+    :param sheet_name: The name of the worksheet to validate.
+    :param unique_togethers: A list of tuples, each tuple contains the column names
+                             that should have combined uniqueness.
+    :return: A list of non-unique-together failures, if any.
+    """
+
+    sheet = workbook[sheet_name]
+    non_unique_together_failures = []
+
+    for columns in unique_togethers:
+        # filter dataframe by these columns and find duplicated rows
+        mask = sheet[list(columns)].duplicated(keep=False)
+
+        if mask.any():
+            failures = [
+                vf.NonUniqueTogetherFailure(sheet=sheet_name, row=tuple(row[1:])) for row in sheet[mask].itertuples()
+            ]
+            non_unique_together_failures.extend(failures)
```

Comment:
Looks good and almost does what we want:

- I think it's a bit more readable if we just use the mask to filter the rows on row 193 rather than in the list comp. We also need to use `drop_duplicates` on the result because we only want one failure pure duplicate (rather than one for each duplicate row)
```
        mask = sheet[list(columns)].duplicated(keep=False)
        duplicated_rows = sheet[mask].drop_duplicates()
```
- we don't need to cast the row to a tuple on 196 because itertuples already return `namedTuple`s.  
- can also increase readability if we set the `index` param in `itertuples` to `False` then we don't have to slice the row

Reply:
done

Reply:
I've added some more comments on the newest review as follow on from this